File: app/src/main/java/com/allever/spycamera/activity/PictureActivity.java
package com.allever.spycamera.activity;

import android.content.Intent;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;

import com.allever.spycamera.RecyclerItemClickListener;
import com.allever.spycamera.adapter.PicAdapter;
import com.allever.spycamera.R;
import com.allever.spycamera.utils.FileUtil;

import java.io.File;
import java.util.List;

/**
 * Created by Allever on 18/5/11.
 */

public class PictureActivity extends AppCompatActivity {
    private static final String TAG = "PictureActivity";
    private List<String> mFilePathList;
    private PicAdapter mAdapter;
    private RecyclerView mRv;
    private String mPkg = "";

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_picture);

        initData();
        initView();
    }

    private void initData(){
        mFilePathList = FileUtil.getAllFilePath();
    }

    private void initView(){
        mRv = findViewById(R.id.id_picture_rv);
        mAdapter = new PicAdapter(this, mFilePathList);
        mRv.setLayoutManager(new GridLayoutManager(this,3));
        mRv.setAdapter(mAdapter);

        mRv.addOnItemTouchListener(new RecyclerItemClickListener(this, mRv, new RecyclerItemClickListener.OnItemClickListener() {
            @Override
            public void onItemClick(View view, int position) {
                if (TextUtils.isEmpty(mPkg)){
                    mPkg = getPkg();
                }
                if (!TextUtils.isEmpty(mPkg)){
                    Intent intent=new Intent();
                    intent.setPackage(mPkg);
                    File file=new File(mFilePathList.get(position));
                    intent.setDataAndType(Uri.fromFile(file),"image/*");
                    intent.setAction(Intent.ACTION_VIEW);
                    startActivity(intent);
                }else {
                    Intent albumIntent = new Intent(Intent.ACTION_VIEW);
                    File file;
                    file = new File(mFilePathList.get(position));
                    albumIntent.setDataAndType(Uri.fromFile(file), "image/*");
                    startActivity(albumIntent);
                }
            }

            @Override
            public void onItemLongClick(View view, int position) {

            }
        }));
    }

    /**
     * 获取带有Launcher应用包名
     * */
    private String getPkg(){
        final Intent mainIntent = new Intent(Intent.ACTION_MAIN, null);
        mainIntent.addCategory(Intent.CATEGORY_LAUNCHER);
        final PackageManager packageManager = getPackageManager();
        List<ResolveInfo> apps = packageManager.queryIntentActivities(mainIntent, 0);
        String pkg = "";
        ResolveInfo info = null;
        for (int i = 0; i < apps.size(); i++) {
            info = apps.get(i);
            Log.e("TAG", info.activityInfo.loadLabel(packageManager) + " pkgName "
                    + info.activityInfo.applicationInfo.packageName + " className " + info.activityInfo.name);

            pkg = info.activityInfo.applicationInfo.packageName;
            if (pkg.contains("gallery") || pkg.contains("photo")){
//                Log.e("TAG", info.activityInfo.loadLabel(packageManager) + " pkgName "
//                    + info.activityInfo.applicationInfo.packageName + " className " + info.activityInfo.name);
                break;
            }
        }
        return pkg;
    }

    /**
     * 获取系统级别的应用包名
     * */
    private String getSystemGalleryPkg(){
        String pkg = "";
        PackageManager packageManager = getPackageManager();
        List<ApplicationInfo> applicationInfoList = packageManager.getInstalledApplications(PackageManager.GET_UNINSTALLED_PACKAGES);
        for (ApplicationInfo app: applicationInfoList){
            if ((app.flags & ApplicationInfo.FLAG_SYSTEM) != 0){
                //系统级应用
                pkg = app.packageName;
                Log.d("TAG", "pkg = " + pkg );
                if (pkg.contains("gallery") || pkg.contains("photo")){
                    Log.d("TAG", "has gallery or photo\n");
                    break;
                }
            }
        }
        return pkg;
    }
}

File: app/src/main/java/com/allever/stealthcamera/ui/PictureActivity.kt
package com.allever.stealthcamera.ui

import android.content.Intent
import android.content.pm.ApplicationInfo
import android.content.pm.PackageManager
import android.content.pm.ResolveInfo
import android.graphics.Rect
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.text.TextUtils
import android.util.Log
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.FileProvider
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.allever.stealthcamera.BuildConfig
import com.allever.stealthcamera.R
import com.allever.stealthcamera.RecyclerItemClickListener
import com.allever.stealthcamera.ui.adapter.PicAdapter
import com.allever.stealthcamera.utils.FileUtil
import java.io.File

/**
 * Created by Allever on 18/5/11.
 */

class PictureActivity : AppCompatActivity() {
    private var mFilePathList: MutableList<String>? = null
    private var mAdapter: PicAdapter? = null
    private var mRv: RecyclerView? = null
    private var mPkg = ""

    /**
     * 获取带有Launcher应用包名
     */
    private val pkg: String
        get() {
            val mainIntent = Intent(Intent.ACTION_MAIN, null)
            mainIntent.addCategory(Intent.CATEGORY_LAUNCHER)
            val packageManager = packageManager
            val apps = packageManager.queryIntentActivities(mainIntent, 0)
            var pkg = ""
            var info: ResolveInfo?
            for (i in apps.indices) {
                info = apps[i]
                Log.e("TAG", info!!.activityInfo.loadLabel(packageManager).toString() + " pkgName "
                        + info.activityInfo.applicationInfo.packageName + " className " + info.activityInfo.name)

                pkg = info.activityInfo.applicationInfo.packageName
                if (pkg.contains("gallery") || pkg.contains("photo")) {
                    break
                }
            }
            return pkg
        }

    /**
     * 获取系统级别的应用包名
     */
    private//系统级应用
    val systemGalleryPkg: String
        get() {
            var pkg = ""
            val packageManager = packageManager
            val applicationInfoList = packageManager.getInstalledApplications(PackageManager.GET_UNINSTALLED_PACKAGES)
            for (app in applicationInfoList) {
                if (app.flags and ApplicationInfo.FLAG_SYSTEM != 0) {
                    pkg = app.packageName
                    Log.d("TAG", "pkg = $pkg")
                    if (pkg.contains("gallery") || pkg.contains("photo")) {
                        Log.d("TAG", "has gallery or photo\n")
                        break
                    }
                }
            }
            return pkg
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_picture)

        initData()
        initView()
    }

    private fun initData() {
        mFilePathList = FileUtil.allFilePath
    }

    private fun initView() {
        mRv = findViewById(R.id.id_picture_rv)
        mAdapter = PicAdapter(this, R.layout.item_picture, mFilePathList!!)
        mRv?.layoutManager = GridLayoutManager(this, 3)
        mRv!!.adapter = mAdapter

//        mRv?.addItemDecoration(object : RecyclerView.ItemDecoration() {
//            override fun getItemOffsets(outRect: Rect, view: View, parent: RecyclerView, state: RecyclerView.State) {
//                val pos = parent.getChildLayoutPosition(view)
//                if (pos == 0) {
//                    outRect.left = firstItemMarginLeft
//                } else if (pos + 1 == mSelectedData.size) {
//                    outRect.right = lastItemMarginRight
//                }
//            }
//        })

        mRv!!.addOnItemTouchListener(RecyclerItemClickListener(this, mRv!!, object : RecyclerItemClickListener.OnItemClickListener {
            override fun onItemClick(view: View, position: Int) {
                if (TextUtils.isEmpty(mPkg)) {
                    mPkg = pkg
                }
                val albumIntent = Intent(Intent.ACTION_VIEW)
                if (!TextUtils.isEmpty(mPkg)) {
                    intent.setPackage(mPkg)
                }

                val file = File(mFilePathList!![position])
                val fileUri: Uri
                fileUri = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                    //解决调用相册不显示图片的问题
                    albumIntent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
                    FileProvider.getUriForFile(
                            this@PictureActivity,
                            BuildConfig.APPLICATION_ID + ".fileprovider",
                            file
                    )
                } else {
                    Uri.fromFile(file)
                }

                albumIntent.setDataAndType(Uri.parse(fileUri.toString()), "image/*")
                startActivity(albumIntent)
            }

            override fun onItemLongClick(view: View?, position: Int) {

            }
        }))
    }

    companion object {
        private val TAG = "PictureActivity"
    }
}
